firebase-middlewares.md

```diff
@@ -0,0 +1,58 @@
+## Using firebase middlewares
+
+In the Front-End, Where ever you are making a Http requests/fetch request like get,post and so on..., the headers need to be attached to these request and its done as below
+
+...
+import {getTokenWithHeaders} from '../../firebase/getTokenWithHeaders'
+
+const header = await getTokenWithHeaders()
+
+await fetch('http://localhost:5000/api/example-end-point', {
+method: "GET/POST",
+headers: header
+})
+.then(res => res.json())
+.then(data => console.log(data))
+.catch(err => console.log(err))
+...
+
+If all these is in place, the token would go to backend
+
+In the Backend, there are middlewares for firebase(authentication.middleware & authorization.middleware) which would be used while creating Endpoints and In your routes folder/ route file it should be done as below
+
+...
+const exampleController = require('../controllers/example.controller');
+const {
+checkIfAuthorized,
+} = require('../lib/middleware/authorization.middleware');
+
+const ROLES = require('../../constants/roles');
+...
+And then,
+
+router.get('/',checkIfAuthorized(ROLES.SUPER_ADMIN),(req, res, next) => {
+exampleController
+.getExample()
+.then((result) => res.json(result))
+.catch(next);
+});
+...
+
+Here there needs to be a check done,while adding middlewares to endpoints, using the below spreadsheet:
+
+Spreadsheet file:
+https://docs.google.com/spreadsheets/d/1FlBJNVvcPEJJpu-RyPz6J-m5Gto7LZ_O90LaXDkJuCs/edit#gid=0
+
+So if there is a "Yes" for the role, it should be used as parameter for middleware,
+
+example: for ADD Batch , we have SuperAdmin & Admin as "YES"
+
+then the middleware should look like:
+
+`checkIfAuthorized(ROLES.SUPER_ADMIN,ROLE.ADMIN)`
+
+example: for Batch Details , we have SuperAdmin & Admin & User as "YES"
+
+then the middleware should look like:
+
+`checkIfAuthorized(ROLES.SUPER_ADMIN,ROLE.ADMIN,ROLE.USER)`
```

Comment:
Shouldn't we use checkIfAuthenticated, because it doesn't matter which role is returning? It only matters that user is authenticated...

Reply:
Yes, you can also use checkIfAutheticated as middleware, if you also need to check the user authentication on backend and no roles needs to be specified :)

Reply:
This is where a claims approach might lessen the confusion. Too late for that so nevermind
Is there a reason for having both an admin and a superadmin?

Reply:
@senner007 I think the roles are defined based on potential future requests, and there is need to check the role only in two cases: add batch and delete batch. Please check the table for more details https://docs.google.com/spreadsheets/d/1FlBJNVvcPEJJpu-RyPz6J-m5Gto7LZ_O90LaXDkJuCs/edit#gid=0 
For everything else IMO, we need only checkIfAuthenticated. 
I would ask for more details about claims approach, but if it is too late for it... then nothing 🙂

Reply:
@cecastosic I see the table yes, but Im asking if theres an actual reason to have both roles?  Cant admins not just be allowed to delete also?
Anyways... should we merge it?

Reply:
If we don't need "Bearer" in front of the token, my answer is yes. 🙂

Reply:
👍 I'll add the changes and merge it

Reply:
@senner008 do you think we only need add checkifAuthorized to Add batch and delete batch and nothing else , based on the attached spreadsheet :)